### PR TITLE
Adding Xinanjiang direct runoff option, changed volume trackers to structure in CFE

### PR DESCRIPTION
**Pull request by jmframe**
[Short description explaining the high-level reason for the pull request]

Adding the Xinanjiang runoff scheme written by Rachel and approved by Fred, addressing issue #10 . Some other issues have been taken care of as well. See the changes.

## Additions

- Xinanjiang direct runoff method available as an option.

## Removals

-

## Changes

- Put Fred's csv parser back in, addressing issue #18 
- Volume trackers are now in a stucture within CFE, this was done by Fred, addressing issue #19 
- Update until is now a simple loop over update, addressing issue #17 , but this might not be the desired fix. I removed some logic from that function.

## Testing

1. Jess's unit tests
2. make and run, when the BMI_CFE reads the forcing data, then runs the model, once with update, and again with update until
3. make and run pass, this tests the bmi ability to get forcings from the bmi functions get_value, set_value
4. make and run pass pet, this tests running two models and passing information between them with BMI
5. All tests done with the XinanJiang runoff option.

## Notes

-

## Todos

- Scott should update the variable_info structure. 
- 

## Checklist

- [X ] PR has an informative and human-readable title
- [ ] Changes are limited to a single goal (no scope creep)
- [ X] Code can be automatically merged (no conflicts)
- [ ] Code follows project standards (link if applicable)
- [ X] Passes all existing automated tests
- [ ] Any _change_ in functionality is tested
- [ ] New functions are documented (with a description, list of inputs, and expected output)
- [ ] Placeholder code is flagged / future todos are captured in comments
- [ ] Visually tested in supported browsers and devices (see checklist below :point_down:)
- [ ] Project documentation has been updated (including the "Unreleased" section of the CHANGELOG)
- [ X] Reviewers requested with the [Reviewers tool](https://help.github.com/articles/requesting-a-pull-request-review/) :arrow_right:

### Target Environment support

- [ ] Windows
- [ ] Linux
- [ ] Browser

### Accessibility

- [ ] Keyboard friendly
- [ ] Screen reader friendly

### Other

- [ ] Is useable without CSS
- [ ] Is useable without JS
- [ ] Flexible from small to large screens
- [ ] No linting errors or warnings
- [ ] JavaScript tests are passing

### CODE REVIEW

**Comment by jmframe** [comment]
@hellkite500 and @robertbartel 

Check out this PR. I modified the update_until to be like the example in the BMI docs (bmi_cfe.c lines 978-1010):

static int Update_until (Bmi *self, double t)
{
    // https://bmi.readthedocs.io/en/latest/#update-until
    // "the time argument can be a non-integral multiple of time steps"

  {
    cfe_state_struct* cfe_ptr = ((cfe_state_struct *) self->data);
    
    double dt;
    double now;

    Get_time_step (self, &dt);
    Get_current_time (self, &now);

    {
      int n;
      double frac;
      const double n_steps = (t - now) / dt;
      for (n=0; n<(int)n_steps; n++) {
        Update (self);
        if (cfe_ptr->verbosity > 1)
            print_cfe_flux_at_timestep(cfe_ptr);

      }

      frac = n_steps - (int)n_steps;
      if (frac > 0)
          printf("WARNING: CFE trying to update a fraction of a timestep\n Figure out what to do here!\n");
    }
  }

  return BMI_SUCCESS;
}

**Comment by hellkite500** [comment]
@madMatchstick can you look at this PR and the previous review comments? Thanks!

File: src/bmi_cfe.c [hunk 1]
```diff
@@ -1641,7 +1677,7 @@ static int Get_state_var_ptrs (Bmi *self, void *ptr_list[])
     ptr_list[54] = &(state->is_forcing_from_bmi );
     ptr_list[55] = state->forcing_file;
     // ####### ptr_list[55] = &(state->forcing_file );
-    ptr_list[56] = &(state->Schaake_adjusted_magic_constant_by_soil_type );
+    ptr_list[56] = &(state->direct_runoff_params_struct.Schaake_adjusted_magic_constant_by_soil_type );
```
[the remark this answers is not in the document]
Comment (madMatchstick) [reply src/bmi_cfe.c:1680]: Will note/apply this to serialization/enhancement dev branches.  Master here should be BMI v2.0 compliment so this function should be removed anyways? @hellkite500

File: src/main.c [hunk 2]
```diff
@@ -54,5 +64,52 @@ int
...
+  double model_time_step_size;
+  cfe_bmi_model2->get_time_step(cfe_bmi_model2, &model_time_step_size);
+  printf("The model time step size is: %lf\n", model_time_step_size);
+
+  printf("looping through and calling update\n");
+  if (cfe_main_data2->verbosity > 0)
+      print_cfe_flux_header();
+
+  // Run the model with the update until function
+  cfe_bmi_model2->update_until(cfe_bmi_model2, 45 * model_time_step_size);
```
[the remark this answers is not in the document]
Comment (madMatchstick) [reply src/main.c:105]: Yes, this seems to be working okay.

File: src/bmi_cfe.c [hunk 3]
```diff
@@ -656,12 +690,33 @@ int read_init_config_cfe(const char* config_file, cfe_state_struct* model, doubl
...
+            return BMI_FAILURE;
+        }   
+    }
 
 #if CFE_DEGUG >= 1
     printf("All CFE config params present\n");
 #endif
 
-    model->Schaake_adjusted_magic_constant_by_soil_type = refkdt * model->NWM_soil_params.satdk / 0.000002;
+    model->direct_runoff_params_struct.Schaake_adjusted_magic_constant_by_soil_type = refkdt * model->NWM_soil_params.satdk / 0.000002;
```
[the remark this answers is not in the document]
Comment (hellkite500) [reply src/bmi_cfe.c:719]: Essentially, yes.  Like I said, it is not super important due to the way things are implemented, but it is a little messy imho.  Just an observation that if only one scheme can be applied at a time, then only one set of parameters ever need get set.

File: src/bmi_cfe.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (hellkite500) [reply src/bmi_cfe.c:1687]: This function doesn't **need** to be removed, it just cannot be added to the BMI struct.

File: src/bmi_cfe.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (madMatchstick) [reply src/bmi_cfe.c]: Finally following... looking better?
```    
if(model->direct_runoff_params_struct.surface_partitioning_scheme == Schaake){
        model->direct_runoff_params_struct.Schaake_adjusted_magic_constant_by_soil_type = refkdt * model->NWM_soil_params.satdk / 0.000002;   
#if CFE_DEGUG >= 1
    printf("Schaake Magic Constant calculated\n");
#endif
    }
```

File: src/bmi_cfe.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (madMatchstick) [reply src/bmi_cfe.c]: As a temp placeholder, replaced with `direct_runoff_params_struct.surface_partitioning_scheme` , i.e.
`typedef enum {Schaake, Xinanjiang} surface_water_partition_type`

To your point, we will need to add remaining optional soil params to appropriate dev/branch...

[the remark this answers is not in the document]
Comment (madMatchstick) [reply src/bmi_cfe.c]: @hellkite500 Are we not removing new BMI functions (serialization) from `bmi.h` here in master?
```
    int (*get_state_var_count)(struct Bmi *self, int *count);
    int (*get_state_var_names)(struct Bmi *self, char ** names);
    int (*get_state_var_types)(struct Bmi *self, char ** types);
    int (*get_state_var_ptrs)(struct Bmi *self, void *ptr_list[]);
    int (*get_state_var_sizes)(struct Bmi *self, unsigned int sizes[]);
    int (*set_state_var)(struct Bmi *self, void *src, int index);
```

[the remark this answers is not in the document]
Comment (hellkite500) [reply src/bmi_cfe.c]: Yes, but that doesn't mean that the DEFINITION of the function can't still exist in the code/library....it just won't be used by BMI.

File: src/bmi_cfe.c [hunk 4]
```diff
@@ -99,7 +99,9 @@ Variable var_info[] = {
 	{ 53, "time_step_size",        "int",  1 },
 	{ 54, "is_forcing_from_bmi",   "int",  1 },
 	{ 55, "forcing_file",                                 "string", 1 },  // strlen
-	{ 56, "Schaake_adjusted_magic_constant_by_soil_type", "double", 1 },
+	//{ 56, "Schaake_adjusted_magic_constant_by_soil_type", "double", 1 },
```
Comment (hellkite500) [inline src/bmi_cfe.c:102]: If this variable still exists, but only conditionally, it probably still needs to exist in this list of all possible variables...

File: src/bmi_cfe.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (madMatchstick) [reply src/bmi_cfe.c]: All `direct_runoff_params_struct` variables now in `info_struc`

File: src/bmi_cfe.c (hunk 4 above)
Comment (madMatchstick) [reply src/bmi_cfe.c:102]: Done. -for all  `direct_runoff_params_struct` vars
```
    { 85, "Schaake_adjusted_magic_constant_by_soil_type",   "double", 1},
    { 86, "a_Xinanjiang_inflection_point_parameter",        "double", 1},
    { 87, "b_Xinanjiang_shape_parameter",                   "double", 1},
    { 88, "x_Xinanjiang_shape_parameter",                   "double", 1}
```

**Review by hellkite500 — APPROVED** [review]

**Pull request merged**